### PR TITLE
Upgrade to ruff 0.2.0 and fix RUF017

### PR DESCRIPTION
**Pull request by Armavica**
* --show-source -> --output-format=full
* --line-length removed from .pre-commit because it is in pyproject
* ruff.extend-per-file-ignores -> ruff.lint.extend-per-file-ignores
* fixed three instances of quadratic list summations

## Related Issue

- [ ] Closes #
- [ ] Related to #

## Checklist

- [x] Checked that [the pre-commit linting/style checks pass](https://docs.pymc.io/en/latest/contributing/python_style.html)
- [ ] Included tests that prove the fix is effective or that the new feature works
- [ ] Added necessary documentation (docstrings and/or example notebooks)
- [x] If you are a pro: each commit corresponds to a [relevant logical change](https://wiki.openstack.org/wiki/GitCommitMessages#Structural_split_of_changes)

## Type of change

- [ ] New feature / enhancement
- [ ] Bug fix
- [ ] Documentation
- [x] Maintenance
- [ ] Other (please specify):

----
📚 Documentation preview 📚: https://pymc--7129.org.readthedocs.build/en/7129/

### CODE REVIEW

**Comment by Armavica** [comment]
It looks like this test failure is unrelated to these changes, as it already occurs in `main`.

File: .pre-commit-config.yaml [hunk 1]
```diff
@@ -24,12 +24,11 @@ repos:
...
         - --exclude=versioneer.py
 - repo: https://github.com/astral-sh/ruff-pre-commit
-  rev: v0.1.14
+  rev: v0.2.0
   hooks:
     - id: ruff
-      args: ["--fix", "--show-source"]
+      args: ["--fix", "--output-format=full"]
     - id: ruff-format
-      args: ["--line-length=100"]
```
Comment (maresb) [inline .pre-commit-config.yaml:32]: Note to self: this is redundant with the setting already in `pyproject.toml`.

Note to self 2: check the commit notes! :joy:

File: requirements-dev.txt [hunk 2]
```diff
@@ -30,4 +30,4 @@ sphinx>=1.5
 sphinxext-rediraffe
 types-cachetools
 typing-extensions>=3.7.4
-watermark
\ No newline at end of file
+watermark
```
Comment (maresb) [inline requirements-dev.txt]: Note: Fixed missing line ending

**Review by maresb — APPROVED** [review]
All good! No changes required, just a few notes to myself below. Thanks @Armavica!

It's nice to fix those sums being used to concatenate lists (!!)

**Comment by maresb** [comment]
Hmm...

```
FAILED tests/variational/test_inference.py::test_fit_start[SVGD-mini] - Failed: DID NOT WARN. No warnings of type (<class 'UserWarning'>,) were emitted.
 Emitted warnings: [].
```

File: pymc/backends/report.py [hunk 3]
```diff
@@ -34,7 +34,7 @@ def __init__(self) -> None:
 
     @property
     def _warnings(self):
-        chains = sum(self._chain_warnings.values(), [])
+        chains = [w for warnings in self._chain_warnings.values() for w in warnings]
```
Comment (ricardoV94) [inline pymc/backends/report.py]: I don't find the two level list comprehension more readable than the sum concatenation tbh. Is this a default code style rule or personal preference?

**Comment by ricardoV94** [comment]
> Hmm...
> 
> ```
> FAILED tests/variational/test_inference.py::test_fit_start[SVGD-mini] - Failed: DID NOT WARN. No warnings of type (<class 'UserWarning'>,) were emitted.
>  Emitted warnings: [].
> ```

it's failing everywhere I'll check Monday

File: pymc/backends/report.py (hunk 3 above)
Comment (ricardoV94) [reply pymc/backends/report.py]: I prefer chain.from_iterable

Comment (Armavica) [reply pymc/backends/report.py]: RUF017 suggests using `ft.reduce`, but the comprehension is faster. But it is indeed less readable, and `it.chain.from_iterable` looks like the best option all around: https://stackoverflow.com/questions/952914/how-do-i-make-a-flat-list-out-of-a-list-of-lists/953097#953097
I'll change everything.

File: pymc/logprob/scan.py [hunk 4]
```diff
@@ -395,19 +395,14 @@ def find_measurable_scans(fgraph, node):
...
-                    # because they only ever have one output
-                    (cl.default_output(), i)
-                    for cl, _ in fgraph.get_clients(out)
-                    if isinstance(cl.op, Subtensor)
-                ]
-                for i, out in enumerate(node.outputs)
-            ],
-            [],
-        )
+        output_clients: list[tuple[Variable, int]] = [
```
Comment (Armavica) [inline pymc/logprob/scan.py:398]: @ricardoV94 I changed the other two, but would you like me to change this one too? It's a bit different

File: pymc/backends/report.py (hunk 3 above)
Comment (maresb) [reply pymc/backends/report.py]: Is it really the double-comprehension that makes this code difficult to read?

Does using itertools actually make this easier to read?

I personally would have used something like: 

```python
warning_groups = self._chain_warnings.values()
chained_warnings = [warning for group in warning_groups for warning in group]
return chained_warnings + self._global_warnings
```

But it's all just style preference.

Comment (ricardoV94) [reply pymc/backends/report.py]: double nested comprehension just to flatten is really odd for me. You are not filtering or doing anything with the items, which is the first thing I would be looking for.

The purpose of the operation is much more obvious with list(chain.from_iterable(iterable))

File: pymc/backends/report.py [hunk 5]
```diff
@@ -13,6 +13,7 @@
 #   limitations under the License.
 
 import dataclasses
+import itertools as it
```
Comment (ricardoV94) [inline pymc/backends/report.py]: just import chain, no need for the it alias?

File: pymc/logprob/scan.py (hunk 4 above)
Comment (ricardoV94) [reply pymc/logprob/scan.py:398]: This one seems fine. The sum was never needed if I'm decoding it correctly?

**Review by ricardoV94 — APPROVED** [review]

File: pymc/backends/report.py (hunk 5 above)
Comment (Armavica) [reply pymc/backends/report.py]: I will do it if you have a strong opinion about it, but I am a bit against extracting functions from their namespaces in general, I think it makes it more difficult to understand where they come from. Especially in the context of PyMC where "chain" can have a whole different meaning…

Comment (maresb) [reply pymc/backends/report.py]: I think the bigger objection here is the alias. How about just `import itertools`?

I think the similarity is giving @ricardoV94 flashbacks to `at` :joy:

Comment (ricardoV94) [reply pymc/backends/report.py]: Yes my problem is the alias

Comment (Armavica) [reply pymc/backends/report.py]: I thought that was a standard alias. I'll remove it.

**Pull request merged**